### PR TITLE
Modifying old RN changes the release time of latest version

### PR DESCRIPTION
**Pull request by Noy-Maimon**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/37694

## Description
Fixed an issue where updating old RNs was modifying the timestamp of the latest RN.

successful build with testing 3 cases tested on Base pack (link for the build [here](https://app.circleci.com/pipelines/github/demisto/content/94798/workflows/e4473d1b-ef4a-4aaf-8a57-f961bdee8e65/jobs/386341)):
1. an old RN version was modified (1.7.0)
image

2. a new RN was added (1.12.11)
image

3. the previous RN to the new RN was modified (1.12.10)
image

another build was triggered for testing only an old RN modification, verifying the latest one's timestamp was not changed. the build was successsfull and can be found [here](https://app.circleci.com/pipelines/github/demisto/content/94803/workflows/cae96936-1750-480f-aff0-d76498a1165f/jobs/386359).

## Minimum version of Cortex XSOAR
- [ ] 5.5.0
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [ ] Tests
- [x] Documentation

### CODE REVIEW

File: Tests/Marketplace/marketplace_services.py [hunk 1]
```diff
@@ -628,7 +628,7 @@ def _load_pack_dependencies(self, index_folder_path, first_level_dependencies, a
         return dependencies_data_result
 
     def _create_changelog_entry(self, release_notes, version_display_name, build_number, pack_was_modified=False,
-                                new_version=True, initial_release=False):
+                                new_version=True, initial_release=False, initial_entry_time=None):
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:631]: I think we should come up with a `update_changelog_entry` function that can get all changed parts of a changelog entry and change them accordingly, we want to keep the build number the same as well.

File: Tests/Marketplace/marketplace_services.py [hunk 2]
```diff
@@ -1079,7 +1084,8 @@ def get_modified_release_notes_lines(self, release_notes_dir: str, changelog_lat
 
         Args:
             release_notes_dir (str): the path to the release notes dir
-            changelog_latest_rn_version (LooseVersion): the last version of release notes in the changelog.json file
+            latest_rn_version (LooseVersion): the last version of release notes in the pack,
+             including the currently changed files
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:1088]: Not sure I understand the description here...

File: Tests/Marketplace/marketplace_services.py [hunk 3]
```diff
@@ -1285,6 +1302,10 @@ def prepare_release_notes(self, index_folder_path, build_number, pack_was_modifi
                         return task_status, not_updated_build
                     else:
                         if latest_release_notes in changelog:
+                            #  Check if the modified files are not only release notes
+                            #  If the modified files are only release notes, don't change the timestamp of the entry
+                            if pack_was_modified:
+                                pack_was_modified = not self.only_release_notes_were_modified(modified_files_paths)
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:1308]: let's move this handling to where that `pack_was_modified` var is first initiated as we want to keep all of this logic in one place.

File: Tests/Marketplace/marketplace_services.py [hunk 4]
```diff
@@ -1236,6 +1242,17 @@ def get_rn_files_names(self, modified_files_paths):
...
+        """
+        Check if all the paths in the modified files paths are of files in the ReleaseNotes directory.
+
+        Args:
+            modified_files_paths (list): list of paths of the pack's modified file
+        Returns:
+            bool: whether the files changed are only ReleaseNotes files
+        """
+        return all(self.RELEASE_NOTES in path for path in modified_files_paths)
+
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:1255]: No need for a whole function for this, just use it as a one-line with a descriptive comment

**Review by guyfreund — CHANGES_REQUESTED** [review]
Looks good 👏 
- Please revert the changes in `Packs/Base/ReleaseNotes/1_12_10.md`.

File: Tests/Marketplace/marketplace_services.py (hunk 3 above)
Comment (Noy-Maimon) [reply Tests/Marketplace/marketplace_services.py:1308]: If we do that, in the case of only modifying RNs, the function _enhance_pack_attributes will not update the released date. Is that desired?

Comment (guyfreund) [reply Tests/Marketplace/marketplace_services.py]: Yes, if only RN files are changed, the version stays the same code-wise, so no need to update the timing. Can you please check if it has other consequences?

File: Tests/Marketplace/marketplace_services.py (hunk 1 above)
Comment (Noy-Maimon) [reply Tests/Marketplace/marketplace_services.py]: Done and tested in this build https://app.circleci.com/pipelines/github/demisto/content/98938/workflows/ec611eac-001e-43db-a773-0f0a500060e2/jobs/394329

File: Tests/Marketplace/marketplace_services.py [hunk 5]
```diff
@@ -629,6 +629,33 @@ def _load_pack_dependencies(self, index_folder_path, first_level_dependencies, a
...
+            changelog (dict): The changelog from the production bucket.
+            version (str): The version that is the key in the changelog of the entry wished to be updated.
+            release_notes (str): The release notes lines to update the entry with.
+            version_display_name (str): The version display name to update the entry with.
+            build_number_with_prefix(srt): the build number to modify the entry to, including the prefix R (if present).
+            released_time: The released time to update the entry with.
+
+        """
+
+        changelog_entry = changelog[version]
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:647]: Let's be more careful here as we might want to reuse this function in the future. Please check whether the version is in the changelog and if not raise an exception.

File: Tests/Marketplace/marketplace_services.py [hunk 6]
```diff
@@ -629,6 +629,33 @@ def _load_pack_dependencies(self, index_folder_path, first_level_dependencies, a
...
+        version_display_name = \
+            version_display_name if version_display_name else changelog_entry['displayName'].split('-')[0]
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:649]: why do we use split and not use it as is?

File: Tests/Marketplace/marketplace_services.py [hunk 7]
```diff
@@ -629,6 +629,33 @@ def _load_pack_dependencies(self, index_folder_path, first_level_dependencies, a
...
+        build_number_with_prefix = \
+            build_number_with_prefix if build_number_with_prefix else changelog_entry['displayName'].split('-')[1]
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:651]: why do we use split and not use it as is?

File: Tests/Marketplace/marketplace_services.py [hunk 8]
```diff
@@ -629,6 +629,33 @@ def _load_pack_dependencies(self, index_folder_path, first_level_dependencies, a
...
+
+        changelog_entry['releaseNotes'] = release_notes if release_notes else changelog_entry['releaseNotes']
+        changelog_entry['displayName'] = f'{version_display_name} - {build_number_with_prefix}'
+        changelog_entry['released'] = released_time if released_time else changelog_entry['released']
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:655]: When developing function in mp_services we try to use global constants that are maintained in `marketplace_constants.py`. Can you please add a new class named `Changelog` and add all these constants? if so please replace the existing ones in `create_changelog_entry` func keep on consistency.

**Review by guyfreund — CHANGES_REQUESTED** [review]
Looking good. Have we checked the new changes?

File: Tests/Marketplace/marketplace_services.py (hunk 6 above)
Comment (Noy-Maimon) [reply Tests/Marketplace/marketplace_services.py:649]: To allow updates on each of these parts separately. I can remove that if you think that is not required.

**Comment by Noy-Maimon** [comment]
@guyfreund all prev changes were tested in this build: https://app.circleci.com/pipelines/github/demisto/content/98938/workflows/ec611eac-001e-43db-a773-0f0a500060e2/jobs/394329

**Review by guyfreund — APPROVED** [review]
I'm approving, please make sure to trigger a build with only a change in an old RN. So that we will be able to see that the timing for the latest version doesn't change. The current build has a change in the latest version of the base pack - therefore we can't test the bugfix.

**Review by guyfreund — APPROVED** [review]

**Comment by Noy-Maimon** [comment]
Final tests build - https://app.circleci.com/pipelines/github/demisto/content/100417/workflows/78430151-b5ab-449b-b2b4-b03cd483645d/jobs/397739

Testing the cases:
1. modifying an old RN
2. creating a new RN + modifying an old RN
3. modifying the last RN

all passed successfully.

**Pull request merged**